pandasql/cost_model.py

```diff
@@ -37,29 +39,42 @@ def _out_of_memory(df, graph):
 
 def _join_then_restrict(df, graph):
     for node in graph.keys():
+        # If result is already cached, no point in looking at it
+        if node.result is not None:
+            continue
+
         # TODO: Look at size of filter/limit for this decision?
         if isinstance(node, ps.core.Selection) or \
                 isinstance(node, ps.core.Limit):
-            join_ancestors = _filter_ancestors(node, graph,
-                                               lambda x:
+            join_ancestors = _filter_ancestors(node, graph, lambda x:
                                                isinstance(x, ps.core.Join))
             if len(join_ancestors) > 0:
                 return True
 
-    return False
-
 
 def _limit_output(df, graph):
-    return isinstance(df, ps.core.Limit)
+    if isinstance(df, ps.core.Limit):
+        return True
 
 
 def _deep_dependency_graph(df, graph):
     ancestors_by_depth = _get_ancestors_by_depth(df, graph)
     depth = len(ancestors_by_depth)
-    return depth > 5
+    if depth > 5:
```

Comment:
Maybe add a TODO to move all the numbers into some sort of configuration object?

Reply:
Will do.